flink-python/pyflink/fn_execution/datastream/process/async_function/queue.py

```diff
@@ -265,9 +275,10 @@ def put(self, windowed_value, timestamp, watermark, record) -> ResultFuture[OUT]
             return entry
 
     def advance_watermark(self, watermark):
-        with self._lock:
-            if watermark > self._current_watermark:
-                self._current_watermark = watermark
+        if watermark > self._current_watermark:
+            self._current_watermark = watermark
+
+            with self._lock:
```

Comment:
`_current_watermark` is only updated when main thread calling  `self._queue.advance_watermark(watermark)` so I think it should be fine to just check and update it here without the `_lock`;
But I am also curious about why we change it here. Do we meet some deadlock case?